engines/sci/graphics/cursor32.cpp

```diff
@@ -0,0 +1,394 @@
+		return;
+	}
+
+	const int16 sourceXOffset = drawRect.left - source.rect.left;
+	const int16 sourceYOffset = drawRect.top - source.rect.top;
+	const int16 drawRectWidth = drawRect.width();
+	const int16 drawRectHeight = drawRect.height();
+
+	byte *targetPixel = target.data + ((drawRect.top - target.rect.top) * target.rect.width()) + (drawRect.left - target.rect.left);
+	const byte *sourcePixel = source.data + (sourceYOffset * source.rect.width()) + sourceXOffset;
+	const uint8 skipColor = source.skipColor;
+
+	const int16 sourceStride = source.rect.width() - drawRectWidth;
+	const int16 targetStride = target.rect.width() - drawRectWidth;
+
+	for (int16 y = 0; y < drawRectHeight; ++y) {
+		for (int16 x = 0; x < drawRectWidth; ++x) {
+			if (*sourcePixel != skipColor) {
+				*targetPixel = *sourcePixel;
+			}
+			++targetPixel;
+			++sourcePixel;
+		}
+		sourcePixel += sourceStride;
+		targetPixel += targetStride;
+	}
+}
+
+void GfxCursor32::drawToHardware(const DrawRegion &source) {
+	Common::Rect drawRect(source.rect);
+	drawRect.clip(_vmapRegion.rect);
+	const int16 sourceXOffset = drawRect.left - source.rect.left;
+	const int16 sourceYOffset = drawRect.top - source.rect.top;
+	byte *sourcePixel = source.data + (sourceYOffset * source.rect.width()) + sourceXOffset;
+
+	g_system->copyRectToScreen(sourcePixel, source.rect.width(), drawRect.left, drawRect.top, drawRect.width(), drawRect.height());
+}
+
+void GfxCursor32::unhide() {
+	if (_hideCount == 0 || --_hideCount) {
+		return;
+	}
+
+	_cursor.rect.moveTo(_position.x - _hotSpot.x, _position.y - _hotSpot.y);
+	revealCursor();
+}
+
+void GfxCursor32::show() {
+	if (_hideCount) {
+		_hideCount = 0;
+		_cursor.rect.moveTo(_position.x - _hotSpot.x, _position.y - _hotSpot.y);
+		revealCursor();
+	}
+}
+
+void GfxCursor32::setRestrictedArea(const Common::Rect &rect) {
+	_restrictedArea = rect;
+
+	const int16 screenWidth = g_sci->_gfxFrameout->getCurrentBuffer().screenWidth;
+	const int16 screenHeight = g_sci->_gfxFrameout->getCurrentBuffer().screenHeight;
+	const int16 scriptWidth = g_sci->_gfxFrameout->getCurrentBuffer().scriptWidth;
+	const int16 scriptHeight = g_sci->_gfxFrameout->getCurrentBuffer().scriptHeight;
+
+	mulru(_restrictedArea, Ratio(screenWidth, scriptWidth), Ratio(screenHeight, scriptHeight), 0);
+
+	if (_position.x < rect.left) {
+		_position.x = rect.left;
+	}
+	if (_position.x >= rect.right) {
+		_position.x = rect.right - 1;
+	}
+	if (_position.y < rect.top) {
+		_position.y = rect.top;
+	}
+	if (_position.y >= rect.bottom) {
+		_position.y = rect.bottom - 1;
+	}
+
+	g_system->warpMouse(_position.x, _position.y);
+}
+
+void GfxCursor32::clearRestrictedArea() {
+	_restrictedArea = _vmapRegion.rect;
+}
+
+void GfxCursor32::setView(const GuiResourceId viewId, const int16 loopNo, const int16 celNo) {
+	hide();
+
+	_cursorInfo.resourceId = viewId;
+	_cursorInfo.loopNo = loopNo;
+	_cursorInfo.celNo = celNo;
+
+	if (viewId != -1) {
+		CelObjView view(viewId, loopNo, celNo);
+
```

Comment:
Ehm, SSCI checked basically nothing at all. That's not acceptable and it's very bad.
Is it a problem to error() out in such case? Or at the very least skip over the code?

SSCI also allowed the scripts to use uninitialized variables. And it didn't check kernel call parameters. SSCI also liked to crash a lot because of all those things. We could have been fine with that, but instead we did it properly and detect such issues. And that's why SCI games, at least the SCI16 ones normally run better in ScummVM than when using the original interpreter. ScummVM crashing is not really acceptable, especially when it's avoidable with a simple error().